src/common.rs

```diff
@@ -319,3 +413,20 @@ impl From<Utf8Error> for GetError {
         GetError
     }
 }
+
+/// Set keys to null where the union member is null.
+///
+/// This is a workaround to https://github.com/apache/arrow-rs/issues/6017#issuecomment-2352756753
+/// - i.e. that dictionary null is most reliably done if the keys are null.
+fn mask_dictionary_keys<K: ArrowPrimitiveType>(keys: &PrimitiveArray<K>, type_ids: &[i8]) -> PrimitiveArray<K> {
+    let mut null_builder = BooleanBufferBuilder::new(keys.len());
+    for key in keys {
+        if key.is_none() || key.is_some_and(|key| type_ids[key.as_usize()] == TYPE_ID_NULL) {
```

Comment:
surely this is less efficient than `if let Some(key) = key {...`?

Reply:
Yep this was just a hangover from hacking to build _something_ 👀